### ISSUE TITLE
Missing CIFAR VOS test option?

### ISSUE THREAD

**Issue by warner-benjamin** [deeplearning-wisc/vos#20]
I was wondering if there is a missing python script for testing VOS trained CIFAR models on the OOD datasets to recreate Table 4 from the paper? In `classification/CIFAR/test.py` I am not seeing any of the VOS code from `train_virtual.py` referenced for any of the evaluation options.

Hopefully I'm not overlooking the code and missing it. Thanks.

**Comment by d12306** [comment]
hi, not sure i unserstand your question, you can use the explicit command line in the readme file for evaluation on classification networks. test.py works fine, you do not need another test_vos.py to do this.

**Comment by warner-benjamin** [comment]
Section 3.3 of the paper states during inference, the output of the logistic regression uncertainty branch is used to predict OOD per equations 8 and 9. But I'm not seeing that in the CIFAR `test.py`. There are options to score MSP, ODIN, Energy, and Mahalanobis in `test.py`, but I am not seeing one for VOS.

In CIFAR `train.virtual.py` lines 139-144 the logistic regression head is setup:
```python
logistic_regression = torch.nn.Linear(1, 2)
logistic_regression = logistic_regression.cuda()
optimizer = torch.optim.SGD(
    list(net.parameters()) + list(weight_energy.parameters()) + \
    list(logistic_regression.parameters()), state['learning_rate'], momentum=state['momentum'],
    weight_decay=state['decay'], nesterov=True)
```
the features for the regression head are extracted via `forward_virtual` on line 192 from `models.wrn_virtual`'s version of WideResNet:
```python
x, output = net.forward_virtual(data)
```
and are later used after the Gaussian sampling for training in lines 248-255 (`weight_energy` via `log_sum_exp`):
```python
energy_score_for_bg = log_sum_exp(predictions_ood, 1)

input_for_lr = torch.cat((energy_score_for_fg, energy_score_for_bg), -1)
labels_for_lr = torch.cat((torch.ones(len(output)).cuda(),
                            torch.zeros(len(ood_samples)).cuda()), -1)

criterion = torch.nn.CrossEntropyLoss()
output1 = logistic_regression(input_for_lr.view(-1, 1))
lr_reg_loss = criterion(output1, labels_for_lr.long())
```
The same code is used in `train_virtual_dense.py` for training DenseNet on CIFAR.

However, in CIFAR's `test.py` I don't see any reference to the logistic regression head, and I am not seeing it used for calculating the uncertainty score for the out of distribution datasets.

In `test.py` WideResNet is imported from `models.wrn_godin` which doesn't have the same `forward_virtual` as `models.wrn_virtual` and the different `forward_virtual` implementation doesn't appear to be used anywhere.

`forward_virtual` is defined the same way as `train_virtual_dense.py`'s DenseNet in `test_gram` but it doesn't appear to be used there during evaluation either.

Thanks for taking the time to answer questions on the code.

**Comment by d12306** [comment]
hi, @warner-benjamin , i got your question. please refer to https://github.com/deeplearning-wisc/vos/issues/14.

The logistic score is just the energy score multiplied by a scalar value in the logistic regression branch. The metrics are essentially the same so you can just use energy score as a metric.

**Comment by warner-benjamin** [comment]
Thanks. CIFAR's test.py currently doesn't use the logistic regression branch at all, but this points me in the right direction.

**Comment by jorism1993** [comment]
I am also having a bit of a hard time to understand this. How did you solve this @warner-benjamin ?
